src/bucket/BucketListSnapshot.cpp

```diff
@@ -41,13 +41,18 @@ BucketListSnapshot::getLedgerSeq() const
     return mLedgerSeq;
 }
 
+// Loops through all buckets in the given snapshot, starting with curr at level
+// 0, then snap at level 0, etc. Calls f on each bucket. Exits early if function
+// returns true
+namespace
+{
 void
-SearchableBucketListSnapshot::loopAllBuckets(
-    std::function<bool(BucketSnapshot const&)> f) const
+loopAllBuckets(std::function<bool(BucketSnapshot const&)> f,
+               std::unique_ptr<BucketListSnapshot const> const& snapshot)
```

Comment:
Ya, `loopAllBuckets` should just take a reference.

However, I think the other instances of taking a pointer reference are correct. The reason being is background threads are the sole owner of their snapshots, but it's the responsibility of the `BucketSnapshotManager` to update those snapshots when it's out of date. This is a little weird, but we wanted to consolidate all the locks and references to `app` in a single connector class (`BucketSnapshotManager`) to limit the scope of synchronization bugs.

The reason we don't just take a reference and modify it is because we want to always create a new snapshot by copy. This is to prevent cases where a buggy update may only partially update a snapshot. The end interface then is passing a reference to the unique_ptr to the `BucketSnapshotManager` who then swaps the pointer to the new snapshot.

Reply:
I'm happy as long as we don't pass const references to non-null unique ptrs around, thanks for updating this.